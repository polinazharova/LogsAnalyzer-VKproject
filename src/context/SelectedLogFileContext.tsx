import React, { createContext, useState } from 'react';


interface SelectedFileContextType {
    selectedFile: string;
    setSelectedFile: React.Dispatch<React.SetStateAction<string>>;
}

const SelectedFileContext = createContext<SelectedFileContextType>({
    selectedFile: "Select a file",
    setSelectedFile: () : void => {},
});


interface SelectedFileProviderProps {
    children: React.ReactNode;
}

const SelectedFileProvider : React.FC<SelectedFileProviderProps> = ({children}) => {

    const [selectedFile, setSelectedFile] = useState<string>("Select a file");

    return (
        <SelectedFileContext.Provider value={{selectedFile, setSelectedFile}}>{children}</SelectedFileContext.Provider>
    )
}

export {SelectedFileContext, SelectedFileProvider};