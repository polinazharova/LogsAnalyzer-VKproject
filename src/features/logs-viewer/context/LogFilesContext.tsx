import React, { createContext, useState } from "react";

interface LogFilesContextType {
  logFiles: File[] | null;
  setLogFiles: React.Dispatch<React.SetStateAction<File[] | null>>;
}
const LogFilesContext = createContext<LogFilesContextType>({
  logFiles: null,
  setLogFiles: (): void => {},
});

interface LogFilesProviderProps {
  children: React.ReactNode;
}
const LogFilesProvider: React.FC<LogFilesProviderProps> = ({ children }) => {
  const [logFiles, setLogFiles] = useState<File[] | null>(null);

  return (
    <LogFilesContext.Provider value={{ logFiles, setLogFiles }}>
      {children}
    </LogFilesContext.Provider>
  );
};

export { LogFilesContext, LogFilesProvider };
