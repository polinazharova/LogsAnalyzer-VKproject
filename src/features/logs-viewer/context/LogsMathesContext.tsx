import React, { createContext, useState } from "react";

interface LogsMatchesContextType {
  logsMatches: { [key: string]: string[] };
  setLogsMatches: React.Dispatch<
    React.SetStateAction<{ [key: string]: string[] }>
  >;
}
const LogsMatchesContext = createContext<LogsMatchesContextType>({
  logsMatches: {},
  setLogsMatches: (): void => {},
});

interface LogsMatchesProviderProps {
  children: React.ReactNode;
}
const LogsMatchesProvider: React.FC<LogsMatchesProviderProps> = ({
  children,
}) => {
  const [logsMatches, setLogsMatches] = useState<{ [key: string]: string[] }>(
    {},
  );

  return (
    <LogsMatchesContext.Provider value={{ logsMatches, setLogsMatches }}>
      {children}
    </LogsMatchesContext.Provider>
  );
};

export { LogsMatchesContext, LogsMatchesProvider };
