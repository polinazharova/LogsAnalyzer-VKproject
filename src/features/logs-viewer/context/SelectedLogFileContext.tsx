import React, { createContext, useState } from "react";

interface SelectedFileContextType {
  selectedFile: string;
  setSelectedFile: React.Dispatch<React.SetStateAction<string>>;
}

const SelectedFileContext = createContext<SelectedFileContextType>({
  selectedFile: "Select a logs file",
  setSelectedFile: (): void => {},
});

interface SelectedFileProviderProps {
  children: React.ReactNode;
}

const SelectedFileProvider: React.FC<SelectedFileProviderProps> = ({
  children,
}) => {
  const [selectedFile, setSelectedFile] =
    useState<string>("Select a logs file");

  return (
    <SelectedFileContext.Provider value={{ selectedFile, setSelectedFile }}>
      {children}
    </SelectedFileContext.Provider>
  );
};

export { SelectedFileContext, SelectedFileProvider };
