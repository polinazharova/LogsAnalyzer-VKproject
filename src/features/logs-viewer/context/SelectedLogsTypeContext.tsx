import React, { createContext, useState } from "react";

interface SelectedLogsTypeContextType {
  selectedLogsType: string;
  setSelectedLogsType: React.Dispatch<React.SetStateAction<string>>;
}
const SelectedLogsTypeContext = createContext<SelectedLogsTypeContextType>({
  selectedLogsType: "Select a log type value",
  setSelectedLogsType: (): void => {},
});

interface SelectedLogsTypeProviderProps {
  children: React.ReactNode;
}
const SelectedLogsTypeProvider: React.FC<SelectedLogsTypeProviderProps> = ({
  children,
}) => {
  const [selectedLogsType, setSelectedLogsType] =
    useState<string>("Select a logs type");

  return (
    <SelectedLogsTypeContext.Provider
      value={{ selectedLogsType, setSelectedLogsType }}
    >
      {children}
    </SelectedLogsTypeContext.Provider>
  );
};

export { SelectedLogsTypeContext, SelectedLogsTypeProvider };
