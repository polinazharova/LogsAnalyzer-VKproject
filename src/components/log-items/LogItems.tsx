import './LogItems.styles.scss'
import { SelectedLogsTypeContext } from '../../context/SelectedLogsTypeContext'
import { useContext, useEffect, useMemo } from 'react';
import LogItem from '../log-item/LogItem';
import { SelectedFileContext } from '../../context/SelectedLogFileContext';

const regExp : RegExp = /^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\+\d{2}:\d{2})\s+\[(\w+)\]\s+(.+?)(\r)?$/;

interface LogItemsProps {
    fileNames: string[];
    logs: {[key: string] : string[]};
    firstShownIndex: number;
    setFirstShownIndex: React.Dispatch<React.SetStateAction<number>>;
    setSize: React.Dispatch<React.SetStateAction<number>>;
};

const LogItems: React.FC<LogItemsProps> = ({fileNames, logs, firstShownIndex, setFirstShownIndex, setSize}) => {
    const {selectedLogsType} = useContext(SelectedLogsTypeContext);
    const {selectedFile} = useContext(SelectedFileContext);

    const { matches, sizeTmp } = useMemo(() => {
        const matches: { [key: string]: string[][] } = {};
        let sizeTmp : number = 0;
      
        fileNames.forEach((fileName) => {
          matches[fileName] = [];
          logs[fileName].forEach((logBody) => {
            const match = logBody.match(regExp);
            if (match) {
              if (
                selectedLogsType !== 'Select a logs type' &&
                selectedLogsType !== 'ALL' &&
                match[2] !== selectedLogsType
              ) {
                return;
              }
              matches[fileName].push([match[1], match[3]]);
              sizeTmp++;
            }
          });
        });
      
        return { matches, sizeTmp };
      }, [logs, selectedLogsType, selectedFile]);

    useEffect(() => {
        setSize(sizeTmp);
        setFirstShownIndex(0);
    }, [logs, selectedLogsType, selectedFile])

    return (
        <div className="main__log-viewer-items">
            {
                fileNames.map((fileName) => (
                    (logs[fileName].length ?
                    (!matches[fileName].length ? <div>Wrong format of logs!</div> : 
                        (matches[fileName].slice(firstShownIndex, firstShownIndex + 10).map((log, index) => (
                        <LogItem key={fileName + index} fileName={fileName} date={log[0]} message={log[1]}/>
                    )))) : <div className="main__log-viewer-item" key={fileName}>{fileName} is empty</div>)
   
                ))
            }
        </div>
    )
}

export default LogItems;
