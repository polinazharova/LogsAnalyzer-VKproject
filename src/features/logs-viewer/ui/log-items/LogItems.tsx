import "./LogItems.styles.scss";
import { SelectedLogsTypeContext } from "../../context/SelectedLogsTypeContext";
import { useContext, useEffect, useMemo } from "react";
import LogItem from "../log-item/LogItem";
import { SelectedFileContext } from "../../context/SelectedLogFileContext";
import useMatches from "../../hooks/useMatches";
import { LogsMatchesContext } from "../../context/LogsMathesContext";

interface LogItemsProps {
  logs: { [key: string]: string[] };
  firstShownIndex: number;
  setFirstShownIndex: React.Dispatch<React.SetStateAction<number>>;
}

const LogItems: React.FC<LogItemsProps> = ({
  logs,
  firstShownIndex,
  setFirstShownIndex,
}) => {
  const { selectedLogsType } = useContext(SelectedLogsTypeContext);
  const { selectedFile } = useContext(SelectedFileContext);

  const fileName = useMemo(() => {
    return Object.keys(logs).includes(selectedFile)
      ? selectedFile
      : Object.keys(logs)[0];
  }, [selectedFile]);

  const { logsMatches, setLogsMatches } = useContext(LogsMatchesContext);

  const matchesTmp :  { [key: string]: string[][] } = useMatches({
    logs,
    fileName,
    selectedLogsType,
    setFirstShownIndex,
  });
  useEffect(() => {
    setLogsMatches(matchesTmp);
  }, [matchesTmp]);

  if (logsMatches[fileName]) {
    return (
      <div className="main__log-viewer-items">
        {logsMatches[fileName].length ? (
          logsMatches[fileName]
            .slice(firstShownIndex, firstShownIndex + 10)
            .map((log: string[], index: number) => (
              <LogItem
                key={fileName + index}
                fileName={fileName}
                date={log[0]}
                message={log[1]}
              />
            ))
        ) : logs[fileName].length ? (
          <div key={`${fileName}-wrong`}>Wrong format of logs!</div>
        ) : (
          <div className="main__log-viewer-item" key={`${fileName}-empty`}>
            {fileName} is empty
          </div>
        )}
      </div>
    );
  }

  return <div>Loading...</div>;
};

export default LogItems;
