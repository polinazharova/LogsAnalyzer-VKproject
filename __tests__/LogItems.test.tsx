import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import LogItems from "../src/features/logs-viewer/ui/log-items/LogItems";
import { SelectedLogsTypeContext } from "../src/features/logs-viewer/context/SelectedLogsTypeContext";
import { SelectedFileContext } from "../src/features/logs-viewer/context/SelectedLogFileContext";
import { LogsMatchesContext } from "../src/features/logs-viewer/context/LogsMathesContext";

const logs = {
  "demo.log": [
    "01.01.1980 [ERROR] Test1",
    "02.01.1980 [INFO] Test2",
    "03.01.1980 [DEBUG] Test3",
    "04.01.1980 [ERROR] Test4",
    "05.01.1980 [ERROR] Test5",
  ],
};

const matches = {
  "demo.log": [
    ["01.01.1980", "Test1"],
    ["02.01.1980", "Test2"],
    ["03.01.1980", "Test3"],
    ["04.01.1980", "Test4"],
    ["05.01.1980", "Test5"],
  ],
};

describe("LogItems component", () => {
  test("renders correctly", async () => {
    render(
      <SelectedLogsTypeContext.Provider
        value={{ selectedLogsType: "ERROR", setSelectedLogsType: jest.fn() }}
      >
        <SelectedFileContext.Provider
          value={{ selectedFile: "demo.log", setSelectedFile: jest.fn() }}
        >
          <LogsMatchesContext.Provider
            value={{ logsMatches: matches, setLogsMatches: jest.fn() }}
          >
            <LogItems
              logs={logs}
              firstShownIndex={0}
              setFirstShownIndex={jest.fn()}
            />
          </LogsMatchesContext.Provider>
        </SelectedFileContext.Provider>
      </SelectedLogsTypeContext.Provider>,
    );

    const container = document.getElementsByClassName(
      "main__log-viewer-items",
    )[0];
    expect(container).toBeInTheDocument();
    const children = document.getElementsByClassName("main__log-viewer-item");
    expect(children.length).toBe(5);
  });
  test("matches snapshot", () => {
    const { container } = render(
      <SelectedLogsTypeContext.Provider
        value={{ selectedLogsType: "ERROR", setSelectedLogsType: jest.fn() }}
      >
        <SelectedFileContext.Provider
          value={{ selectedFile: "demo.log", setSelectedFile: jest.fn() }}
        >
          <LogsMatchesContext.Provider
            value={{ logsMatches: matches, setLogsMatches: jest.fn() }}
          >
            <LogItems
              logs={logs}
              firstShownIndex={0}
              setFirstShownIndex={jest.fn()}
            />
          </LogsMatchesContext.Provider>
        </SelectedFileContext.Provider>
      </SelectedLogsTypeContext.Provider>,
    );
    expect(container).not.toBeEmptyDOMElement();
    expect(container.firstChild).toMatchSnapshot();
  });
  test("does not show log items when logs len === 0", () => {
    render(
      <SelectedLogsTypeContext.Provider
        value={{ selectedLogsType: "ERROR", setSelectedLogsType: jest.fn() }}
      >
        <SelectedFileContext.Provider
          value={{ selectedFile: "demo.log", setSelectedFile: jest.fn() }}
        >
          <LogsMatchesContext.Provider
            value={{
              logsMatches: { "demo.txt": [] },
              setLogsMatches: jest.fn(),
            }}
          >
            <LogItems
              logs={{ "demo.txt": [] }}
              firstShownIndex={0}
              setFirstShownIndex={jest.fn()}
            />
          </LogsMatchesContext.Provider>
        </SelectedFileContext.Provider>
      </SelectedLogsTypeContext.Provider>,
    );

    expect(screen.getByText("demo.txt is empty")).toBeInTheDocument();
  });

  test("does not show log items when matches len === 0", () => {
    render(
      <SelectedLogsTypeContext.Provider
        value={{ selectedLogsType: "ERROR", setSelectedLogsType: jest.fn() }}
      >
        <SelectedFileContext.Provider
          value={{ selectedFile: "demo.log", setSelectedFile: jest.fn() }}
        >
          <LogsMatchesContext.Provider
            value={{
              logsMatches: { "demo.txt": [] },
              setLogsMatches: jest.fn(),
            }}
          >
            <LogItems
              logs={{ "demo.txt": ["01.01.1980 [INFO] Test"] }}
              firstShownIndex={0}
              setFirstShownIndex={jest.fn()}
            />
          </LogsMatchesContext.Provider>
        </SelectedFileContext.Provider>
      </SelectedLogsTypeContext.Provider>,
    );

    expect(
      screen.getByText(
        `Wrong format of logs or logs with selected type don't exist!`,
      ),
    ).toBeInTheDocument();
  });
});
