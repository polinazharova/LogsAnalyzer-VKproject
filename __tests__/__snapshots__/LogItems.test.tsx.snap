// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogItems component matches snapshot 1`] = `
<div
  class="main__log-viewer-items"
>
  <div
    class="main__log-viewer-item"
  >
    <img
      alt="file icon from flaticon.com"
      class="main__log-viewer-item-file-icon"
      src=""
    />
    <span
      class="main__log-viewer-item-filename"
    >
      demo.log
    </span>
    <span
      class="main__log-viewer-item-log-date"
    >
      01.01.1980, 00:00:00
    </span>
    <span
      class="main__log-viewer-item-log-message"
    >
      Test1
    </span>
  </div>
  <div
    class="main__log-viewer-item"
  >
    <img
      alt="file icon from flaticon.com"
      class="main__log-viewer-item-file-icon"
      src=""
    />
    <span
      class="main__log-viewer-item-filename"
    >
      demo.log
    </span>
    <span
      class="main__log-viewer-item-log-date"
    >
      01.02.1980, 00:00:00
    </span>
    <span
      class="main__log-viewer-item-log-message"
    >
      Test2
    </span>
  </div>
  <div
    class="main__log-viewer-item"
  >
    <img
      alt="file icon from flaticon.com"
      class="main__log-viewer-item-file-icon"
      src=""
    />
    <span
      class="main__log-viewer-item-filename"
    >
      demo.log
    </span>
    <span
      class="main__log-viewer-item-log-date"
    >
      01.03.1980, 00:00:00
    </span>
    <span
      class="main__log-viewer-item-log-message"
    >
      Test3
    </span>
  </div>
  <div
    class="main__log-viewer-item"
  >
    <img
      alt="file icon from flaticon.com"
      class="main__log-viewer-item-file-icon"
      src=""
    />
    <span
      class="main__log-viewer-item-filename"
    >
      demo.log
    </span>
    <span
      class="main__log-viewer-item-log-date"
    >
      01.04.1980, 00:00:00
    </span>
    <span
      class="main__log-viewer-item-log-message"
    >
      Test4
    </span>
  </div>
  <div
    class="main__log-viewer-item"
  >
    <img
      alt="file icon from flaticon.com"
      class="main__log-viewer-item-file-icon"
      src=""
    />
    <span
      class="main__log-viewer-item-filename"
    >
      demo.log
    </span>
    <span
      class="main__log-viewer-item-log-date"
    >
      01.05.1980, 00:00:00
    </span>
    <span
      class="main__log-viewer-item-log-message"
    >
      Test5
    </span>
  </div>
</div>
`;
